tensorflow/core/kernels/eigen_backward_spatial_convolutions.h

```diff
@@ -312,71 +312,33 @@ SpatialConvolutionBackwardInput(
-                                const Eigen::array<
-                                    typename internal::traits<Input>::Index,
-                                    internal::traits<Input>::NumDimensions>,
-                                const Input>>>,
-                    const TensorReshapingOp<
-                        const DSizes<typename internal::traits<Input>::Index,
-                                     2>,
-                        const TensorImagePatchOp<
-                            Dynamic, Dynamic,
-                            const Eigen::TensorForcedEvalOp<
-                                const Eigen::TensorShufflingOp<
-                                    const Eigen::array<
-                                        typename internal::traits<Input>::Index,
-                                        internal::traits<Input>::NumDimensions>,
-                                    const OutputBackward>>>>>>>>>,
-    const TensorReverseOp<
-        const Eigen::array<typename internal::traits<Input>::Index,
-                           internal::traits<Input>::NumDimensions>,
-        const Eigen::TensorForcedEvalOp<const Eigen::TensorShufflingOp<
-            const Eigen::array<typename internal::traits<Input>::Index,
-                               internal::traits<Input>::NumDimensions>,
-            const Eigen::TensorReshapingOp<
-                const Eigen::DSizes<typename internal::traits<Input>::Index,
-                                    internal::traits<Input>::NumDimensions>,
-                const TensorContractionOp<
-                    const array<
-                        IndexPair<typename internal::traits<Input>::Index>, 1>,
-                    const TensorReshapingOp<
-                        const DSizes<typename internal::traits<Input>::Index,
-                                     2>,
-                        const TensorImagePatchOp<
-                            Dynamic, Dynamic,
-                            const Eigen::TensorForcedEvalOp<
-                                const Eigen::TensorShufflingOp<
-                                    const Eigen::array<
-                                        typename internal::traits<Input>::Index,
-                                        internal::traits<Input>::NumDimensions>,
-                                    const OutputBackward>>>>,
-                    const TensorReshapingOp<
-                        const DSizes<typename internal::traits<Input>::Index,
-                                     2>,
-                        const Eigen::TensorForcedEvalOp<
-                            const Eigen::TensorShufflingOp<
-                                const Eigen::array<
-                                    typename internal::traits<Input>::Index,
-                                    internal::traits<Input>::NumDimensions>,
-                                const Input>>>>>>>>>
+    internal::traits<OutputBackward>::Layout == ColMajor,
```

Comment:
Wow, this return type is nasty.  I think we can just use `auto` these days?  I believe the original kernel came from a time when we didn't allow c++14 auto return types.

Reply:
I completely agree, auto works absolutely fine under Cpp20, but I didn't want to change the style of the code too much, as I thought that the "template hell" return types are there for a reason. If it were up to me, the return type would be auto everywhere nowadays, because it allows much quicker changes, as modifying a single expression in the body doesn't require rewriting the whole templated type.